### PR TITLE
AN-6745: Migration between app versions freezes the app - Part 1.

### PR DESCRIPTION
**Pull request by BradleyWire**
## What's new in this PR?

JIRA: https://wearezeta.atlassian.net/browse/AN-6745

### Issues

When you update the application from 3.46 to 3.47 it'll freeze the application due to throwing an exception around unhandled migrations. 

### Causes

The Schema for the Scala database (SQLiteDatabase) doesn't explicitly define "NON NULL" modifiers to columns, so when the room creates a "room_master_table" for each of the tables it's mimicking it doesn't correspond to the mandatory NON NULL values that come with @PrimaryKey annotation (amongst other things).  

### Solutions

To migrate all tables in the database with the exact schema as it was before but with the explicit NON-NULL modifiers attached. Any breaking schema modifications will need to be done the next migration away from **127** and **25**.  

## Notes

All migrations will be Unit Tested for every version to ensure the data remains in tact. 
#### APK
[Download build #1518](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1518/artifact/build/artifact/wire-dev-PR2679-1518.apk)
[Download build #1519](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1519/artifact/build/artifact/wire-dev-PR2679-1519.apk)
[Download build #1522](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1522/artifact/build/artifact/wire-dev-PR2679-1522.apk)
[Download build #1526](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1526/artifact/build/artifact/wire-dev-PR2679-1526.apk)
[Download build #1527](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1527/artifact/build/artifact/wire-dev-PR2679-1527.apk)

### CODE REVIEW

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/sqlite/GlobalSQLiteDbTestHelper.kt [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+                    null,
+                    contentValues,
+                    SQLiteDatabase.CONFLICT_REPLACE
+                )
+                close()
+            }
+        }
+
+
+        fun createTable(testOpenHelper: GlobalDbSQLiteOpenHelper) {
```
Comment (gizemb) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/sqlite/GlobalSQLiteDbTestHelper.kt]: this looks like a duplicate of GlobalDbSQLiteOpenHelper's onCreate() method

Comment (BradleyWire) [reply storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/sqlite/GlobalSQLiteDbTestHelper.kt]: Ah yes, I changed the way I tested it throughout. The onCreate became redundant. Good catch!

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/IntegrationTest.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import androidx.room.testing.MigrationTestHelper
+import androidx.sqlite.db.SupportSQLiteDatabase
+import androidx.sqlite.db.framework.FrameworkSQLiteOpenHelperFactory
+import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner
+import androidx.test.platform.app.InstrumentationRegistry
+import org.junit.Rule
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4ClassRunner::class)
+abstract class IntegrationTest(assetLocation: String?) {
```
Comment (gizemb) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/IntegrationTest.kt]: maybe we can separate Room testing parts from general integration tests

Comment (BradleyWire) [reply storage/src/androidTest/kotlin/com/waz/zclient/storage/IntegrationTest.kt]: What do you mean? Use composition for the `testHelper`?

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            registeredPush = TEST_ACTIVE_ACCOUNT_REGISTERED_PUSH,
+            openHelper = testOpenHelper
+        )
+
+        val db = validateMigraton()
+
+        runBlocking {
+            val activeAccounts = getActiveAccounts()
+            val account = activeAccounts[0]
+            assert(account.id == TEST_ACTIVE_ACCOUNT_ID)
```
Comment (mejdoo) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt]: What about using Kluent in all the Kotlin unit tests?

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            val teams = getTeams()
+            val team = teams[0]
+            assert(team.teamId == TEST_TEAM_ID)
+            assert(team.creatorId == TEST_TEAM_CREATOR)
+            assert(team.iconId == TEST_TEAM_ICON)
+            assert(team.teamName == TEST_TEAM_NAME)
+        }
+    }
+
+    private fun validateMigraton() =
```
Comment (mejdoo) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt]: typo `validateMigration`

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/sqlite/GlobalSQLiteDbTestHelper.kt [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+    override fun onDowngrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
+        //Not required for testing version 24 to 25
+    }
+
+}
+
+class GlobalSQLiteDbTestHelper private constructor() {
+
+    companion object {
```
Comment (mejdoo) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/sqlite/GlobalSQLiteDbTestHelper.kt]: move `companion object` to the bottom

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt [hunk 6]
```diff
@@ -52,8 +56,6 @@ class GlobalDatabase24to25MigrationTest : IntegrationTest(GlobalDatabase::class.
             assert(account.accessToken?.expiresInMillis == 1582896705028)
             assert(account.refreshToken == TEST_ACTIVE_ACCOUNT_COOKIE)
         }
-
-        closeDb(db)
```
Comment (gizemb) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt]: don't we need these anymore?

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/MigrationTestHelper.kt [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        MigrationTestHelper(
+            InstrumentationRegistry.getInstrumentation(),
+            assetLocation,
+            FrameworkSQLiteOpenHelperFactory()
+        )
+
+    fun validateMigration(
+        dbName: String,
+        dbVersion: Int,
+        validateDroppedTabled: Boolean,
```
Comment (gizemb) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/MigrationTestHelper.kt]: there's a small typo here but not really important

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/sqlite/GlobalSQLiteDbTestHelper.kt (hunk 5 above)
Comment (BradleyWire) [reply storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/sqlite/GlobalSQLiteDbTestHelper.kt]: The companion object is the only object in this class, it is at the bottom in essence.

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt (hunk 3 above)
Comment (BradleyWire) [reply storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt]: Doesn't work properly. I throws an exception and fails all the tests.

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt (hunk 6 above)
Comment (BradleyWire) [reply storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt]: No we don't. I think a better use-case would be for room to room migrations by the looks of it.

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt (hunk 3 above)
Comment (mejdoo) [reply storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt]: if it's not stable we shouldn't use it

Comment (BradleyWire) [reply storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt]: Bring it to the collective I reckon

Comment (BradleyWire) [reply storage/src/androidTest/kotlin/com/waz/zclient/storage/db/accountdata/GlobalDatabase24to25MigrationTest.kt]:
```
com.android.builder.dexing.DexArchiveBuilderException: com.android.builder.dexing.DexArchiveBuilderException: Failed to process /Users/BradleyWilson/.gradle/caches/modules-2/files-2.1/org.amshove.kluent/kluent/1.60/29f3e927a3a5011de2eeb72ba759efd65a8762ca/kluent-1.60.jar
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)

```

This is the exception for reference.

**Pull request merged**